### ISSUE TITLE
Performance degradation in Keeper

### ISSUE THREAD

**Issue by tavplubix** [ClickHouse/ClickHouse#37494]
I tried to start `clickhouse-server` after pulling and building fresh master and 13 minutes were not enough for Keeper to startup:

```
2022.05.24 20:32:50.423649 [ 1234691 ] {} <Trace> KeeperLogStore: Continue to write into /home/tavplubix/ch/test/repl/coordination/logs/changelog_1_100000.bin.zstd
2022.05.24 20:45:31.912196 [ 1234691 ] {} <Information> KeeperServer: No config in snapshot, will use config from log store with log index 87542
...
2022.05.24 20:46:32.336150 [ 1234691 ] {} <Error> void DB::KeeperDispatcher::initialize(const Poco::Util::AbstractConfiguration &, bool, bool): Code: 568. DB::Exception: Failed to wait RAFT initialization. (RAFT_ERROR), Stack trace (when copying this message, always include the lines below):
...
2022.05.24 20:46:32.351183 [ 1234691 ] {} <Debug> KeeperDispatcher: Shutting down storage dispatcher
2022.05.24 20:46:32.357640 [ 1234691 ] {} <Information> RaftInstance: shutting down raft core
2022.05.24 20:46:32.357748 [ 1234691 ] {} <Information> RaftInstance: sent stop signal to the commit thread.
2022.05.24 20:46:32.357811 [ 1234691 ] {} <Information> RaftInstance: cancelled all schedulers.
2022.05.24 20:50:47.209627 [ 1235282 ] {} <Information> RaftInstance: config at index 16990 is committed, prev config log idx 87542
...
2022.05.24 20:50:48.003885 [ 1235282 ] {} <Debug> KeeperStateMachine: Session ID response 14 with timeout 30000
2022.05.24 20:54:51.010968 [ 1235282 ] {} <Information> RaftInstance: config at index 29624 is committed, prev config log idx 87542
2022.05.24 20:54:51.011097 [ 1235282 ] {} <Information> RaftInstance: skipped config 29624, latest config 87542
2022.05.24 20:54:51.011224 [ 1235282 ] {} <Debug> KeeperStateMachine: Session ID response 15 with timeout 30000
2022.05.24 20:56:45.943733 [ 1235282 ] {} <Information> RaftInstance: config at index 36486 is committed, prev config log idx 87542
...
```
Also you can see freezes on shutdown.
(Almost) full log: https://pastila.nl/?001c5a5c/b91589b1223fbf8f92ad93bfe356ef32
Content of `coordination` dir: https://drive.google.com/file/d/1I888dCliGc4v531rPdfikM6Ur4zHjrwJ/view?usp=sharing
Some stacktraces: https://pastila.nl/?000ef5aa/62787cf94935c6b84007360462e1e6b3 (the most interesting is `Thread 1`)

Also 100 seconds is not enough to shutdown KeeperDispatcher after stress tests sometimes:
Report: https://s3.amazonaws.com/clickhouse-test-reports/0/b9bea7d80355e24da17a19abcbada98af6b319ef/stress_test__memory__actions_.html
Some stacktraces: https://pastila.nl/?002530a6/6e5d041dd91124b4c3c630a8c2b3ecbd

Notice that in the first case it was debug build and in the second case it was build with msan (and probably ThreadFuzzer was active). Of course it does not make sense to measure performance in such scenarios. But there is no way it can be _that_ slow. And we should check if release build slowed down too or not (probably it did).

Probably related to #37036, cc: @antonio2368

**Comment by antonio2368** [comment]
The startup performance should be fixed with #37252 
Initial startup takes a couple of seconds with your logs and the entire sync with a second node takes 1 minute on a Debug build on my machine (I will check out the Release build, but it should be much better).

I'll check out the problem with shutdown.

**Comment by tavplubix** [comment]
Similar issue with slow shutdown: https://s3.amazonaws.com/clickhouse-test-reports/0/8b4ec600636361a7955dfb3c177a7b2c6b37c61b/stress_test__debug__actions_.html

However, it happened in BC check (I haven't seen this in master lately). Do we need to backport something to 22.6?

**Comment by tavplubix** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/0/57a719bafdb23fb955540859fd76b81efad36d18/stress_test__debug__actions_.html

**Comment by antonio2368** [comment]
> Similar issue with slow shutdown: https://s3.amazonaws.com/clickhouse-test-reports/0/8b4ec600636361a7955dfb3c177a7b2c6b37c61b/stress_test__debug__actions_.html
> 
> However, it happened in BC check (I haven't seen this in master lately). Do we need to backport something to 22.6?

Nothing was merged in 22.6 that could cause a performance degradation (nothing I'm aware of...)
